src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/filesystem/BitbucketSCMFileSystem.java

```diff
@@ -0,0 +1,159 @@
+        this.ref = ref;
+    }
+
+    @Override
+    public SCMFile getRoot() {
+        return new BitbucketSCMFile(client, ref);
+    }
+
+    // We do not provide this information in the REST response, so this is undefined.
+    @Override
+    public long lastModified() {
+        return 0L;
+    }
+
+    @Extension
+    public static class BuilderImpl extends SCMFileSystem.Builder {
+
+        @Inject
+        BitbucketClientFactoryProvider clientFactoryProvider;
+        @Inject
+        JenkinsToBitbucketCredentials jenkinsToBitbucketCredentials;
+        @Inject
+        BitbucketPluginConfiguration pluginConfiguration;
+
+        @Override
+        public SCMFileSystem build(Item item, SCM scm,
+                                   @CheckForNull SCMRevision scmRevision) {
+            if (!(scm instanceof BitbucketSCM)) {
+                return null;
+            }
+
+            BitbucketSCM bitbucketSCM = (BitbucketSCM) scm;
+            Optional<BitbucketServerConfiguration> maybeServerConfiguration =
+                    pluginConfiguration.getServerById(bitbucketSCM.getServerId());
+            if (!maybeServerConfiguration.isPresent() || maybeServerConfiguration.get().validate().kind == Kind.ERROR) {
+                LOGGER.finer("ERROR: Bitbucket Server configuration for job " + item.getName() +
+                             " is invalid- cannot build file system");
+                return null;
+            }
+
+            BitbucketSCMRepository repository = bitbucketSCM.getBitbucketSCMRepository();
+
+            BitbucketFilePathClient filePathClient =
+                    clientFactoryProvider.getClient(maybeServerConfiguration.get().getBaseUrl(),
+                            jenkinsToBitbucketCredentials.toBitbucketCredentials(repository.getCredentialsId()))
+                            .getProjectClient(repository.getProjectKey())
+                            .getRepositoryClient(repository.getRepositorySlug())
+                            .getFilePathClient();
+
+            return new BitbucketSCMFileSystem(filePathClient, null, bitbucketSCM.getBranches().get(0).toString());
+        }
+
+        @Override
+        public SCMFileSystem build(SCMSource source, SCMHead head,
+                                   @CheckForNull SCMRevision scmRevision) {
+            if (!(source instanceof BitbucketSCMSource)) {
+                return null;
+            }
+            BitbucketSCMSource bitbucketSCMSource = (BitbucketSCMSource) source;
+            String ownerName = source.getOwner() == null ? "" : source.getOwner().getName();
+            Optional<BitbucketServerConfiguration> maybeServerConfiguration =
+                    pluginConfiguration.getServerById(bitbucketSCMSource.getServerId());
+            if (!maybeServerConfiguration.isPresent() || maybeServerConfiguration.get().validate().kind == Kind.ERROR) {
+                LOGGER.finer("ERROR: Bitbucket Server configuration for job " + ownerName +
```

Comment:
Maybe this should be `LOGGER.warning` instead? It seems like its a bit more important than `finer`. I'm unfamiliar with this logging system though - please correct me if i'm wrong about this

Reply:
I went with finer instead of warning because I'm a bit worried about noise- a job can still run and complete even if the lightweight checkout fails, and in this specific case (missing/invalid server configuration), we will output proper logging when the job fails anyway. The only other case I can think of is where someone uses a different job type with a `GitSCMHead` type we don't know how to handle- particularly for branch sources that could be quite noisy as every branch will go through this so I think finer logging makes more sense.
But I'm happy to change if you feel strongly about it.

Reply:
I should note I wanted to go with job logging- where WARN would make perfect sense but alas, there's no way to access the task listener from inside the SCMFileSystem.

Reply:
Revisiting this I completely disagree with Martin from 5 weeks ago. Job config is borked, a warning is completely appropriate.